crates/flox-rust-sdk/src/models/publish.rs

```diff
@@ -244,13 +236,132 @@ impl<'flox> Publish<'flox, NixAnalysis> {
+    ) -> Result<Self, PublishError> {
+        let url = publish_ref.clone_url();
+        let repo_dir = tempfile::tempdir_in(temp_dir).unwrap().into_path(); // todo catch error
+        let repo = <Git as GitProvider>::clone(&url, &repo_dir, false).await?;
+
+        Ok(Self(repo))
+    }
+
+    fn catalog_branch_name(system: &System) -> String {
+        format!("catalog/{system}")
+    }
+
+    /// Create an [UpstreamCatalog] by checking out or creating a catalog branch.
+    ///
+    /// `Git` objects can switch branches at any time leaving the repo in an unknown state.
+    /// [get_catalog] ensures that only one [UpstreamCatalog] exists at a time by requiring a `&mut self`.
+    async fn get_catalog(&mut self, system: &System) -> Result<UpstreamCatalog, PublishError> {
+        if self.0.list_branches().await? // todo: catch error
+            .into_iter().any(|info| info.name == Self::catalog_branch_name(system))
+        {
+            self.0
+                .checkout(&Self::catalog_branch_name(system), false)
+                .await?; // todo: catch error
+        } else {
+            self.0
+                .checkout(&Self::catalog_branch_name(system), true)
+                .await?;
+            self.0
+                .set_origin(&Self::catalog_branch_name(system), "origin")
+                .await?;
+        }
+        Ok(UpstreamCatalog(&self.0))
+    }
+}
+
+/// Representation of a specific catalog branch in an exclusive clone of an upstream repo.
+///
+/// [UpstreamCatalog] guaranteesd that during its lifetime all operations on the underlying git repo
+/// are performed on a single branch,
+/// and that the branch is pushed to upstream before a new branch can be checked out.
+struct UpstreamCatalog<'a>(&'a Git);
+
+impl UpstreamCatalog<'_> {
+    /// Mostly naïve approxiaton of a snapshot path
+    /// TODO: fix before releasing this PR!
+    fn get_snapshot_path(&self, snapshot: &Value) -> PathBuf {
+        let mut path = self
+            .0
+            .workdir()
+            .unwrap()
+            .join("packages")
+            .join(snapshot["eval"]["meta"]["pname"].to_string());
+        snapshot["eval"]["meta"]["version"].to_string();
+
+        path.set_file_name(format!("{}.json", snapshot["eval"]["meta"]["version"]));
+        path
+    }
+
+    /// Try retrieving a snapshot from the catalog
+    /// TODO: better addressing (attrpath + version + drv hash?)
+    fn get_snapshot(&self, snapshot: &Value) -> Result<Option<Value>, PublishError> {
+        let path = self.get_snapshot_path(snapshot);
+        let read_snapshot = match fs::read_to_string(path) {
+            Ok(s) => Some(serde_json::from_str(&s)?),
+            Err(e) if e.kind() == io::ErrorKind::NotFound => None,
+            Err(e) => Err(e)?,
+        };
+
+        Ok(read_snapshot)
+    }
+
+    /// [Value] to be [flox_types::catalog::CatalogEntry]
+    ///
+    /// Consumers should check if a snapshot already exists with [Self::get_snapshot]
```

Comment:
Why not perform the check here? Also no caller is using the return value of `get_snapshot`, so it could be a `contains_snapshot() -> bool` method instead

Reply:
I didn't want to restrict behaviour around republishing/overwriting at this level that should be a decision of the Application level.

`get_snapshot` returns a snapshot so we can do updates on them rather than blindly overriding. -- I do concede, none of that is implemented and our application check would likely be a `get_snapshot()?.is_some()`.